src/rust/engine/dep_inference/src/javascript/mod.rs

```diff
@@ -18,42 +23,115 @@ include!(concat!(env!("OUT_DIR"), "/javascript/constants.rs"));
 include!(concat!(env!("OUT_DIR"), "/javascript/visitor.rs"));
 include!(concat!(env!("OUT_DIR"), "/javascript_impl_hash.rs"));
 
-#[derive(Serialize, Deserialize)]
+#[derive(Debug, Serialize, Deserialize)]
 pub struct ParsedJavascriptDependencies {
+    pub imports: HashMap<String, JavascriptImportInfo>,
+}
+
+#[derive(Debug, PartialEq, Serialize, Deserialize)]
+pub struct JavascriptImportInfo {
     pub file_imports: HashSet<String>,
     pub package_imports: HashSet<String>,
 }
 
+impl JavascriptImportInfo {
+    fn from_sets(file_imports: HashSet<String>, package_imports: HashSet<String>) -> Self {
+        JavascriptImportInfo {
+            file_imports,
+            package_imports,
+        }
+    }
+
+    #[cfg(test)]
+    fn new<F: AsRef<str>, T: AsRef<str>>(
+        file_imports: impl IntoIterator<Item = F>,
+        package_imports: impl IntoIterator<Item = T>,
+    ) -> Self {
+        JavascriptImportInfo::from_sets(
+            file_imports
+                .into_iter()
+                .map(|s| s.as_ref().to_string())
+                .collect(),
+            package_imports
+                .into_iter()
+                .map(|s| s.as_ref().to_string())
+                .collect(),
+        )
+    }
+}
+
+fn patterns_as_lookup(patterns: Vec<ImportPattern>) -> HashMap<String, Vec<String>> {
+    patterns
+        .into_iter()
+        .map(|pattern| (pattern.pattern, pattern.replacements))
+        .collect()
+}
+
+fn placed_at_root(package_root: &str, string: &str) -> bool {
+    !package_root.is_empty() && string.starts_with(package_root)
+}
+
+fn is_relative_specifier(string: &str) -> bool {
+    ['.', '/'].into_iter().any(|p| string.starts_with(p))
```

Comment:
Minor rust thing: `starts_with` is generic over a "pattern", and thus supports directly asking "starts with any of these":
```suggestion
    string.starts_with(&['.', '/'])
```

See docs: https://doc.rust-lang.org/std/primitive.str.html#method.starts_with

Reply:
I kinda just assumed a slice of chars would be tested as sequence, not an or. My mistake!

Reply:
Yeah, "starts with a sequence" is also supported, but with a str arg: `string.starts_with("abc")` matches `abcdef` but not `bcdef`.